### PR TITLE
Check module import validity with eslint-plugin-import

### PR DESCRIPTION
**Pull request by jsnmoon**
(Assists #11688.)

This change adds [eslint-plugin-import](https://github.com/benmosher/eslint-plugin-import) to our ESLint configuration. It uses either a [Webpack resolver](https://github.com/benmosher/eslint-plugin-import/tree/master/resolvers/webpack) or a [Node resolver](https://github.com/benmosher/eslint-plugin-import/tree/master/resolvers/node) depending on the path of the linted file.

This change helps us verify the validity of existing imports as well as ensuring correctness for any new imports we introduce into the project. This will be especially important as we begin migrating the client module notation from CommonJS to ES.

**Todo List**:
- [x]  Correctly parse imports inside the `/client/` directory
- [x]  Correctly parse imports inside the `/server/` directory
- [x]  Correctly parse imports inside a `/test/` directory

CC @ockham @aduth @ehg: Your input would be greatly appreciated!

### CODE REVIEW

**Comment by nosolosw** [comment]
Howdy Jason, regarding the changes in `.eslintrc`: with the merge of https://github.com/Automattic/wp-calypso/pull/6945 you may set any new rules as errors. All new code will be reported as an `error` if it doesn't honor the rule, but the legacy code will be reported as `warning`, so our CI/hook tools won't complain about it.

**Comment by jsnmoon** [comment]
Hey Andrés, thanks for the heads up! That's good to know :)

File: webpack.config.js [hunk 1]
```diff
@@ -64,15 +65,7 @@ const webpackConfig = {
 			}
 		]
 	},
-	resolve: {
```
Comment (aduth) [inline webpack.config.js]: Does this need to be moved to a separate config? Wouldn't it be enough to resolve using the `webpack.config.js` file from `.eslint-resolver.js` as-is?

File: .eslint-resolver.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+const webpackResolverConfig = {
+	config: {
+		resolve: webpackResolveConfig,
+	},
+};
+
+exports.interfaceVersion = 2
+
+exports.resolve = function (source, file, config) {
```
Comment (aduth) [inline .eslint-resolver.js]: Ironically your ESLint resolver script has many lint errors 😄  (spacing after function, between parens).

File: .eslint-resolver.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+ * Internal dependencies
+ */
+const webpackResolveConfig = require( './webpack.config.resolve' );
+
+/**
+ * Module variables
+ */
+const log = debugFactory('eslint-plugin-import:resolver:wp-calypso-resolver');
+
+const nodeResolverConfig = {
```
Comment (aduth) [inline .eslint-resolver.js]: Can you explain the need for separate resolving configs? Should this be pulled from `webpack.config.node.js` for server-side module importing?

File: .eslintrc.js [hunk 4]
```diff
@@ -1,3 +1,8 @@
+/* eslint-disable import/no-commonjs */
+
+const path = require( 'path' );
+const customResolverPath = path.join( __dirname, './.eslint-resolver.js' );
```
Comment (aduth) [inline .eslintrc.js]: Won't the result of this join be simply "./.eslint-resolver.js" which we could instead define as a string key of the `settings`, not a computed property?

File: .eslintrc.js [hunk 5]
```diff
@@ -10,11 +15,32 @@ module.exports = {
...
+	],
+	settings: {
+		'import/resolver': {
+			[ customResolverPath ]: {}
+		}
+	},
 	rules: {
 		camelcase: 0, // REST API objects include underscores
+		// NOTE: Some import rules are errors in client (Webpack's resolution) and
+		//       warnings in server (Node's module resolution).
```
Comment (aduth) [inline .eslintrc.js]: Can't we still define these all as errors and expect the server-specific configuration to override them?

File: index.js [hunk 6]
```diff
@@ -1,4 +1,5 @@
 /* eslint-disable no-console */
+/* eslint-disable import/no-commonjs */
```
Comment (aduth) [inline index.js]: FYI you can comma-separate disabled rules in a single `eslint-disable` comment.

File: .eslintrc.js (hunk 4 above)
Comment (jsnmoon) [inline .eslintrc.js]: Yes, you're absolutely correct. I'll address this

File: .eslint-resolver.js (hunk 2 above)
Comment (jsnmoon) [inline .eslint-resolver.js]: D'oh! 😅

File: index.js (hunk 6 above)
Comment (jsnmoon) [inline index.js]: Got it 👍

File: webpack.config.js (hunk 1 above)
Comment (jsnmoon) [inline webpack.config.js]: Hi @aduth, the main reason I had separated the `resolve` into its own separate file was due to the fact that `webpack.config.js` has `lib/create-config` in its import path. 

Easy repro: `node -e "var config = require('./webpack.config');"` (execute this at project root)

There doesn't seem to be an easy way to modify the node path for ESLint - what do you suggest?

File: .eslint-resolver.js (hunk 2 above)
Comment (jsnmoon) [inline .eslint-resolver.js]: I realized that ESLint was ignoring the file due to the `.eslint` prefix. I went ahead and changed `.eslintignore` to force parsing `.eslint-resolver` so this shouldn't happen again.

File: .eslintrc.js (hunk 5 above)
Comment (jsnmoon) [inline .eslintrc.js]: Yes, we can define all of these rules as errors and override them inside `server/.eslintrc.json`. The following rules are already overridden for the server: 
- Set to warn:
  - `import/default`
  - `import/export`
  - `import/named`
  - `import/namespace`
- Set to ignore:
  -` import/no-commonjs`

Which rules would you recommend upgrading to error in `.eslintrc.js`? The following rules are currently set as warnings for the client: 
- `import/no-commonjs`
- `import/no-named-as-default-member`
- `import/no-named-as-default`
- `import/no-webpack-loader-syntax`

File: .eslint-resolver.js (hunk 3 above)
Comment (jsnmoon) [inline .eslint-resolver.js]: Sure - the reason why we need separate resolving configs is due to how files across our project assume a different root directory. For client files, this includes `/client/`
and `/client/extensions/`; for server files, this includes `/`, `/client/`, and `/server/`. Test files tend to add `/test/` as a root directory as well.

I do think it should be pulling from `webpack.config.node.js` for server-side module importing. I'll push out a revision to address this shortly.

**Comment by jsnmoon** [comment]
Given that we're beginning to rely more in Prettier, I think we can put this PR on hold indefinitely for now. I'll leave the branch alive for anyone who'd like to use this as a reference to integrate import path resolution with ESLint.

**Pull request closed without merging**